aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -58,13 +58,66 @@ static inline Tensor& unary_op_impl_(Tensor& self, OutImpl& out_impl) {
   return out_impl(self, self);
 }
 
-Tensor& acos_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, acos_stub); }
-Tensor acos(const Tensor& self) { return unary_op_impl(self, at::acos_out); }
-Tensor& acos_(Tensor& self) { return unary_op_impl_(self, at::acos_out); }
+// unary_floating_ufunc_* functions are helper functions to support integral and bool Tensors
+// to default floating type promotion for floating ufuncs. For non-floating ufuncs, use existing
+// helper functions (unary_op_impl*) if needed.
+template <typename Stub>
+static inline Tensor& unary_floating_ufunc_op_impl_out(Tensor& result, const Tensor& self, Stub& stub) {
+  TORCH_CHECK(!isIntegralType(result.scalar_type(), /*includeBool=*/ true),
+    "Attempted to call a function that returns a tensor of floating dtype ",
+    "but requested a tensor of bool or integral dtype. ");
+
+  auto iter = TensorIterator::unary_floating_ufunc(result, self,
+    /*check_mem_overlap=*/true);
+  stub(iter.device_type(), iter);
+  return result;
+}
+
+template <typename OutImpl>
+static inline Tensor unary_floating_ufunc_op_impl(const Tensor& self, OutImpl& out_impl) {
+  if (isIntegralType(self.scalar_type(), /*includeBool=*/ true)) {
+    const auto scalar_type = typeMetaToScalarType(c10::get_default_dtype());
+    Tensor result = at::empty({0}, self.options().dtype(scalar_type));
+    return out_impl(result, self);
+  }
+
+  Tensor result = at::empty({0}, self.options());
+  return out_impl(result, self);
+}
+
+template <typename OutImpl>
+static inline Tensor unary_floating_ufunc_cast_op_impl(
+    const Tensor& self,
+    OutImpl& out_impl) {
+  if (isIntegralType(self.scalar_type(), /*includeBool=*/ true)) {
+    const auto scalar_type = typeMetaToScalarType(c10::get_default_dtype());
+    Tensor result = at::empty({0}, self.options().dtype(scalar_type));
+    return out_impl(result, self.to(scalar_type));
```

Comment:
out of the loop for a bit fighting with CI for autocasting... @mruberry git blame (or credit!) says it's your work (from a [merged branch](https://github.com/pytorch/pytorch/commit/fa0ea9a49a1353ac996f88ab624591c520a91a61)) to split `unary...impl` into two versions.  The other version (`unary_floating_ufunc_op_impl`), for integer tensor input, creates result as the default type and sends self and result along to TensorIterator as they are.  This version precasts self to match the output type, so TensorIterator doesn't need to handle mismatched types internally.  

Splitting those two versions is a good way to deal with the fact that some backend functions may handle mismatched self+result types and some may not.  Afaict `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_VEC` macros are the only place `unary_floating_ufunc_cast_op_impl` is used, ie, the ops handled by those macros are the ones you decided should receive precasting.

I'd like to make sure I understand how you and @krshrimali decided which ops receive precasting.  The ops handled by `IMPLEMENT_UNARY_FLOATING_UFUNC_OP_VEC` match the ops from [this](https://github.com/pytorch/pytorch/pull/33322#discussion_r394318659) (possibly outdated) comment, with the addition of `erfinv` and `lgamma`.  If i'm reading that comment correctly, the ops listed there don't support internal int->default_type promotion when called on CUDA tensors, which is a reasonable criterion to route them through `unary_floating_ufunc_cast_op_impl` rather than `unary_floating_ufunc_op_impl`.

However (if im reading it correctly) those same ops _do_ support internal promote when called with CPU inputs.  Right now the `_impl` each op calls is based only on the op itself.  Since internal promote support depends on both the op and `self`'s device, do you think it makes sense to have `unary_floating_ufunc_cast_op_impl` choose not to precast if `is_cuda()` is false?

side note:  `unary_floating_ufunc_cast_op_impl` only precasts integer inputs, in other words, floating point inputs never precast no matter which `_impl` they route through, and always rely on internal TensorIterator promotion.  That's great for GPU performance, and will be great for performance with the hoped-for output `dtype=...` argument as well.

Reply:
I agree to what you suggest here @mcarilli. For the functions using the helper macros, adding a simple if-else condition works fine with all the tests:

```cpp
#define IMPLEMENT_UNARY_FLOATING_UFUNC_OP_CORE(op)                     \
  Tensor op(const Tensor& self) {                                      \
    if (!self.is_cuda) \
      return unary_floating_ufunc_op_impl(self, at::op##_out); \
    else \
      return unary_floating_ufunc_cast_op_impl(self, at::op##_out);      \
  }
```

@mruberry - What are your views?